Interface-Based Architecture

0x471 opened the pull request:
## Description
This PR implements interface-based architecture to improve developer experience while maintaining full backward compatibility. This addresses the complexity of the current Fungible Token Contract by providing both simple and advanced usage patterns with interfaces. Additionally integrates configs restructuring, sideloaded interfaces, and admin method improvements.

## Changes
* Implemented a wrapper (CoreFungibleToken.ts) which only has main fungible token functionalities
* Implemented TokenConfig interface for reading token configuration settings
* Implemented TokenAdmin interface for admin-only operations such as updating the side-loaded verification key hash, updating configs and more
* Implemented SideloadedTokenContract interface for token operations like mint, burn and more with side-loaded proofs.

## Meeting Follow-Up
* Removed CoreFungibleToken wrapper based on review feedback
* Converted interface files to `.d.ts` format (TokenAdmin.ts → admin.d.ts, SideloadedTokenContract.ts → sideloaded.d.ts)
* Created core.d.ts interface for standard token operations (mint, burn, transfer, getBalance, etc.)
* Renamed interfaces for consistency: TokenAdmin → Admin, SideloadedTokenContract → Sideloaded
* FungibleToken now implements all three interfaces (Admin, Core, Sideloaded)
* Added JSDoc comments to interface files and removed them from FungibleTokenContract

## Merged
- https://github.com/o1-labs-XT/fungible-token-contract/pull/41
- https://github.com/o1-labs-XT/fungible-token-contract/pull/42
- https://github.com/o1-labs-XT/fungible-token-contract/pull/43

0x471 committed:
implement interface-based architecture for improved developer experience

0x471 committed:
move src/configs.ts to lib directory

0x471 committed:
standardize side-loaded proof interfaces and relocate examples

0x471 committed:
add getAdmin() method to retrieve fungible token admin

45930 commented on src/interfaces/SideloadedTokenContract.ts:
You should be able to use the `.d.ts` file extension which means "type declaration" file.  Then it will fail to compile if any javascript execution code is written in these files.  It's a good way to protect the purpose of these files in the future.

45930 commented on src/CoreFungibleToken.ts:
```diff
@@ -0,0 +1,119 @@
+  maxAmount: UInt64.MAXINT(),
+});
+
+export class CoreToken {
```
Can we change one of the examples to use this class, or just annotate a simple example in typedoc comments?

Fizzixnerd commented on src/CoreFungibleToken.ts:
```diff
@@ -0,0 +1,119 @@
+  maxAmount: UInt64.MAXINT(),
+});
+
+export class CoreToken {
```
I'm not totally against the interface pattern, however.  It's useful for grouping related functionality.  It's just, have the main `FungibleToken` implement each of the interfaces.  Wouldn't that be cleaner?  Then, if you only need `TokenAdmin` privileges, your function will just accept something that implements the TokenAdmin interface.  That will likely be the FungibleToken itself, but it allows people to use the type system to be more sure they aren't calling methods they don't mean to.

0x471 committed:
remove CoreFungibleToken wrapper

0x471 committed:
convert interfaces to `.d.ts` format and rename for consistency

0x471 committed:
update implementation to use new interface names

0x471 commented:
Based on the meeting we had yesterday, I've updated this PR to implement a cleaner interface-based architecture approach.

Here are the key changes made:
- Removed the `CoreFungibleToken` wrapper
- Converted interface files to `.d.ts` format
- Besides `sideloaded.d.ts` and `admin.d.ts` I also created `core.d.ts` for rest of the (not internal) operations.
- Renamed interfaces for consistency: `TokenAdmin` → `Admin`, `SideloadedTokenContract` → `Sideloaded`
- `FungibleToken` now implements all three interfaces (`Admin`, `Core`, `Sideloaded`)
- Added comprehensive JSDoc comments to the interface files and removed them from the `FungibleTokenContract`

Tests pass with `proofsEnabled=true`

0x471 committed:
Merge pull request #43 from o1-labs-XT/feature/add-get-admin-method

Add `getAdmin()` Method to Retrieve Fungible Token Admin

0x471 committed:
Merge pull request #42 from o1-labs-XT/refactor/standardize-sideloaded-interfaces

Standardize Side-Loaded Proof Interfaces and Relocate Examples

0x471 committed:
Merge remote-tracking branch 'origin/refactor/move-configs-to-lib' into refactor/interface-organization
